### ISSUE TITLE
AMD 9950x Unable to cpu info and cannot be installed

### ISSUE THREAD

**Issue by organics2016** [microsoft/DeepSpeed#6625]
It doesn't seem to work on new CPUs

# Install info
```
C:\Users\xxxxxxx>pip install deepspeed
Collecting deepspeed
  Using cached deepspeed-0.15.2.tar.gz (1.4 MB)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [28 lines of output]
      Traceback (most recent call last):
        File "C:\Users\organics\AppData\Local\Temp\pip-install-87s7bsm4\deepspeed_901174b213be4815bb532f3a2bf07d1f\op_builder\builder.py", line 398, in cpu_arch
          from cpuinfo import get_cpu_info
      ModuleNotFoundError: No module named 'cpuinfo'

      During handling of the above exception, another exception occurred:

      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "C:\Users\organics\AppData\Local\Temp\pip-install-87s7bsm4\deepspeed_901174b213be4815bb532f3a2bf07d1f\setup.py", line 200, in <module>
          ext_modules.append(builder.builder())
        File "C:\Users\organics\AppData\Local\Temp\pip-install-87s7bsm4\deepspeed_901174b213be4815bb532f3a2bf07d1f\op_builder\builder.py", line 699, in builder
          {'cxx': self.strip_empty_entries(self.cxx_args()), \
        File "C:\Users\organics\AppData\Local\Temp\pip-install-87s7bsm4\deepspeed_901174b213be4815bb532f3a2bf07d1f\op_builder\builder.py", line 840, in cxx_args
          CPU_ARCH = self.cpu_arch()
        File "C:\Users\organics\AppData\Local\Temp\pip-install-87s7bsm4\deepspeed_901174b213be4815bb532f3a2bf07d1f\op_builder\builder.py", line 400, in cpu_arch
          cpu_info = self._backup_cpuinfo()
        File "C:\Users\organics\AppData\Local\Temp\pip-install-87s7bsm4\deepspeed_901174b213be4815bb532f3a2bf07d1f\op_builder\builder.py", line 430, in _backup_cpuinfo
          if not self.command_exists('lscpu'):
        File "C:\Users\organics\AppData\Local\Temp\pip-install-87s7bsm4\deepspeed_901174b213be4815bb532f3a2bf07d1f\op_builder\builder.py", line 486, in command_exists
          result = subprocess.Popen(safe_cmd, stdout=subprocess.PIPE)
        File "C:\Users\organics\miniconda3\envs\kanojyo2\lib\subprocess.py", line 971, in __init__
          self._execute_child(args, executable, preexec_fn, close_fds,
        File "C:\Users\organics\miniconda3\envs\kanojyo2\lib\subprocess.py", line 1456, in _execute_child
          hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
      FileNotFoundError: [WinError 2] 系统找不到指定的文件。
      DS_BUILD_OPS=1
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

# CPU info
```
	Name			AMD Ryzen 9
	Codename		Granite Ridge
	Specification		AMD Ryzen 9 9950X 16-Core Processor            
	Package 		
	CPUID			F.4.0
	Extended CPUID		1A.44
	Core Stepping		
	Technology		4 nm
	Core Speed		5470.5 MHz
	Multiplier x Bus Speed	54.86 x 99.7 MHz
	Base frequency (cores)	99.7 MHz
	Base frequency (mem.)	99.7 MHz
	Instructions sets	MMX (+), SSE, SSE2, SSE3, SSSE3, SSE4.1, SSE4.2, SSE4A, x86-64, AES, AVX, AVX2, AVX512 (DQ, BW, VL, CD, IFMA, VBMI, VBMI2, VNNI, BITALG, VPOPCNTDQ, VP2INTERSECT, BF16), FMA3, SHA
	Microcode Revision	0xB404022
```

# System info
```
Windows 11 IoT LTSC
24H2
2024/‎10/‎8
26100.2033
Windows Feature Experience Pack 1000.26100.23.0
```

# Pytorch
`2.4.1+cu121`

# CUDA
```
nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2023 NVIDIA Corporation
Built on Wed_Feb__8_05:53:42_Coordinated_Universal_Time_2023
Cuda compilation tools, release 12.1, V12.1.66
Build cuda_12.1.r12.1/compiler.32415258_0
```

**Comment by jomayeri** [comment]
Can you run `pip install py-cpuinfo` and rerun?

**Comment by organics2016** [comment]
> Can you run `pip install py-cpuinfo` and rerun?

Thanks for reply，this is my problem, but DeepSpeed seems to be difficult to install on Windows.

This is installed after `pip install py-cpuinfo`

```
(deepspeed) C:\Users\organics>pip install deepspeed
Collecting deepspeed
  Using cached deepspeed-0.15.2.tar.gz (1.4 MB)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [18 lines of output]
      [2024-10-19 12:13:21,630] [INFO] [real_accelerator.py:219:get_accelerator] Setting ds_accelerator to cuda (auto detect)
      [2024-10-19 12:13:22,455] [INFO] [real_accelerator.py:219:get_accelerator] Setting ds_accelerator to cuda (auto detect)
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "C:\Users\organics\AppData\Local\Temp\pip-install-l0n9d9pp\deepspeed_b2c2a539534a401da05463798adefafa\setup.py", line 200, in <module>
          ext_modules.append(builder.builder())
        File "C:\Users\organics\AppData\Local\Temp\pip-install-l0n9d9pp\deepspeed_b2c2a539534a401da05463798adefafa\op_builder\builder.py", line 720, in builder
          extra_link_args=self.strip_empty_entries(self.extra_ldflags()))
        File "C:\Users\organics\AppData\Local\Temp\pip-install-l0n9d9pp\deepspeed_b2c2a539534a401da05463798adefafa\op_builder\inference_cutlass_builder.py", line 74, in extra_ldflags
          import dskernels
      ModuleNotFoundError: No module named 'dskernels'
      DS_BUILD_OPS=1
       [WARNING]  Skip pre-compile of incompatible async_io; One can disable async_io with DS_BUILD_AIO=0
       [WARNING]  Skip pre-compile of incompatible evoformer_attn; One can disable evoformer_attn with DS_BUILD_EVOFORMER_ATTN=0
       [WARNING]  Skip pre-compile of incompatible fp_quantizer; One can disable fp_quantizer with DS_BUILD_FP_QUANTIZER=0
       [WARNING]  Skip pre-compile of incompatible gds; One can disable gds with DS_BUILD_GDS=0
       [WARNING]  Filtered compute capabilities ['6.0', '6.1', '7.0']
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

I try to it
```
(deepspeed) C:\Users\organics>set DS_BUILD_AIO=0
(deepspeed) C:\Users\organics>set DS_BUILD_FP_QUANTIZER=0
(deepspeed) C:\Users\organics>set DS_BUILD_GDS=0
(deepspeed) C:\Users\organics>set DS_BUILD_SPARSE_ATTN=0

(deepspeed) C:\Users\organics>pip install deepspeed
Collecting deepspeed
  Using cached deepspeed-0.15.2.tar.gz (1.4 MB)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [18 lines of output]
      [2024-10-19 12:26:04,947] [INFO] [real_accelerator.py:219:get_accelerator] Setting ds_accelerator to cuda (auto detect)
      [2024-10-19 12:26:05,787] [INFO] [real_accelerator.py:219:get_accelerator] Setting ds_accelerator to cuda (auto detect)
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "C:\Users\organics\AppData\Local\Temp\pip-install-xdx7h744\deepspeed_d6481dcc53cc4154a7ee5af867caf4f6\setup.py", line 209, in <module>
          result = subprocess.check_output(git_hash_cmd)
        File "C:\Users\organics\miniconda3\envs\deepspeed\lib\subprocess.py", line 421, in check_output
          return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
        File "C:\Users\organics\miniconda3\envs\deepspeed\lib\subprocess.py", line 503, in run
          with Popen(*popenargs, **kwargs) as process:
        File "C:\Users\organics\miniconda3\envs\deepspeed\lib\subprocess.py", line 971, in __init__
          self._execute_child(args, executable, preexec_fn, close_fds,
        File "C:\Users\organics\miniconda3\envs\deepspeed\lib\subprocess.py", line 1456, in _execute_child
          hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
      FileNotFoundError: [WinError 2] 系统找不到指定的文件。
      DS_BUILD_OPS=1
      Install Ops={'async_io': False, 'fused_adam': 1, 'cpu_adam': 1, 'cpu_adagrad': 1, 'cpu_lion': 1, 'evoformer_attn': False, 'fp_quantizer': False, 'fused_lamb': 1, 'fused_lion': 1, 'gds': False, 'transformer_inference': 1, 'inference_core_ops': 1, 'cutlass_ops': False, 'quantizer': 1, 'ragged_device_ops': False, 'ragged_ops': 1, 'random_ltd': 1, 'sparse_attn': False, 'spatial_inference': 1, 'transformer': 1, 'stochastic_transformer': 1}
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

**Comment by xiezhipeng-git** [comment]
PS D:\my\work\study\ai\DeepSpeed> python setup.py install
[2024-10-27 01:02:29,527] [INFO] [real_accelerator.py:219:get_accelerator] Setting ds_accelerator to cuda (auto detect)
Traceback (most recent call last):
  File "D:\my\work\study\ai\DeepSpeed\setup.py", line 40, in <module>
    from op_builder import get_default_compute_capabilities, OpBuilder
  File "D:\my\work\study\ai\DeepSpeed\op_builder\__init__.py", line 18, in <module>
    import deepspeed.ops.op_builder  # noqa: F401 # type: ignore
  File "D:\my\work\study\ai\DeepSpeed\deepspeed\__init__.py", line 25, in <module>
    from . import ops
  File "D:\my\work\study\ai\DeepSpeed\deepspeed\ops\__init__.py", line 11, in <module>
    from . import transformer
  File "D:\my\work\study\ai\DeepSpeed\deepspeed\ops\transformer\__init__.py", line 7, in <module>
    from .inference.config import DeepSpeedInferenceConfig
  File "D:\my\work\study\ai\DeepSpeed\deepspeed\ops\transformer\inference\__init__.py", line 7, in <module>
    from ....model_implementations.transformers.ds_transformer import DeepSpeedTransformerInference
  File "D:\my\work\study\ai\DeepSpeed\deepspeed\model_implementations\__init__.py", line 6, in <module>
    from .transformers.ds_transformer import DeepSpeedTransformerInference
  File "D:\my\work\study\ai\DeepSpeed\deepspeed\model_implementations\transformers\ds_transformer.py", line 18, in <module>
    from deepspeed.ops.transformer.inference.triton.mlp import TritonMLP
  File "D:\my\work\study\ai\DeepSpeed\deepspeed\ops\transformer\inference\triton\__init__.py", line 10, in <module>
    from .ops import *
  File "D:\my\work\study\ai\DeepSpeed\deepspeed\ops\transformer\inference\triton\ops.py", line 6, in <module>
    import deepspeed.ops.transformer.inference.triton.matmul_ext as matmul_ext
  File "D:\my\work\study\ai\DeepSpeed\deepspeed\ops\transformer\inference\triton\matmul_ext.py", line 461, in <module>
    fp16_matmul = Fp16Matmul()
  File "D:\my\work\study\ai\DeepSpeed\deepspeed\ops\transformer\inference\triton\matmul_ext.py", line 210, in __init__
    __class__._read_autotune_table()
  File "D:\my\work\study\ai\DeepSpeed\deepspeed\ops\transformer\inference\triton\matmul_ext.py", line 444, in _read_autotune_table
    TritonMatmul._read_autotune_table(__class__.__name__ + "_2d_kernel", __class__._2d_kernel)
  File "D:\my\work\study\ai\DeepSpeed\deepspeed\ops\transformer\inference\triton\matmul_ext.py", line 165, in _read_autotune_table
    cache_manager = AutotuneCacheManager(cache_key)
  File "D:\my\work\study\ai\DeepSpeed\deepspeed\ops\transformer\inference\triton\matmul_ext.py", line 87, in __init__
    TritonCacheDir.warn_if_nfs(self.cache_dir)
  File "D:\my\work\study\ai\DeepSpeed\deepspeed\ops\transformer\inference\triton\matmul_ext.py", line 44, in warn_if_nfs
    if is_nfs_path(cache_dir) and not TritonCacheDir._warning_printed:
  File "D:\my\work\study\ai\DeepSpeed\deepspeed\ops\transformer\inference\triton\matmul_ext.py", line 27, in is_nfs_path
    output = subprocess.check_output(['df', '-T', path], encoding='utf-8')
  File "D:\my\env\python3.10.10\lib\subprocess.py", line 421, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "D:\my\env\python3.10.10\lib\subprocess.py", line 503, in run
    with Popen(*popenargs, **kwargs) as process:
  File "D:\my\env\python3.10.10\lib\subprocess.py", line 971, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "D:\my\env\python3.10.10\lib\subprocess.py", line 1440, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
FileNotFoundError: [WinError 2] 系统找不到指定的文件。
I have a similar problem.

**Comment by xiezhipeng-git** [comment]
need change code in deepspeed\__init__.py file 
```python
from . import ops
from . import module_inject
```
to
```python
from deepspeed import ops
from deepspeed import module_inject
```
then 
change deepspeed\ops\transformer\inference\triton\matmul_ext.py
```python
def is_nfs_path(path):
    # Normalize the path to get the absolute path
    path = os.path.abspath(path)

    if platform.system() == "Windows":
        # 在 Windows 上不执行 df 命令，直接返回 False
        return False
    else:
        # Use the 'df' command to find the file system type for the given path
        try:
            output = subprocess.check_output(['df', '-T', path], encoding='utf-8')
        except subprocess.CalledProcessError:
            return False  # Command failed

        # Process the output of 'df -T' to check for 'nfs' in the filesystem type column
        lines = output.strip().split('\n')
        if len(lines) > 1:  # The first line is headers
            fs_type = lines[1].split()[1].lower()  # File system type is the second column
            return 'nfs' in fs_type
        return False

```
then new error
Using d:\my\env\python3.10.10\lib\site-packages
Finished processing dependencies for deepspeed==0.15.4+unknown
deepspeed build time = 14.176020860671997 secs
Exception ignored in atexit callback: <function matmul_ext_update_autotune_table at 0x00000199AAC5E830>
Traceback (most recent call last):
  File "D:\my\work\study\ai\DeepSpeed\deepspeed\ops\transformer\inference\triton\matmul_ext.py", line 481, in matmul_ext_update_autotune_table
    fp16_matmul._update_autotune_table()
  File "D:\my\work\study\ai\DeepSpeed\deepspeed\ops\transformer\inference\triton\matmul_ext.py", line 458, in _update_autotune_table
    TritonMatmul._update_autotune_table(__class__.__name__ + "_2d_kernel", __class__._2d_kernel)
  File "D:\my\work\study\ai\DeepSpeed\deepspeed\ops\transformer\inference\triton\matmul_ext.py", line 187, in _update_autotune_table
    cache_manager.put(autotune_table)
  File "D:\my\work\study\ai\DeepSpeed\deepspeed\ops\transformer\inference\triton\matmul_ext.py", line 106, in put
    os.rename(self.file_path + ".tmp", self.file_path)
FileExistsError: [WinError 183] 当文件已存在时，无法创建该文件。: 'C:\\Users\\Admin\\.triton\\autotune\\Fp16Matmul_2d_kernel.pickle.tmp' -> 'C:\\Users\\Admin\\.triton\\autotune\\Fp16Matmul_2d_kernel.pickle'

**Comment by organics2016** [comment]
@loadams Thanks for your reply.
Perfect installation
```
pip install https://files.pythonhosted.org/packages/2e/5c/2058713749655a6b1830ecb8d7db61637a396239610aeb11b59974d29b66/deepspeed-0.15.0-cp311-cp311-win_amd64.whl
........
Successfully installed MarkupSafe-3.0.2 annotated-types-0.7.0 colorama-0.4.6 deepspeed-0.15.0 filelock-3.16.1 fsspec-2024.10.0 hjson-3.1.0 jinja2-3.1.4 mpmath-1.3.0 networkx-3.4.2 ninja-1.11.1.1 numpy-2.1.2 nvidia-ml-py-12.560.30 packaging-24.1 psutil-6.1.0 py-cpuinfo-9.0.0 pydantic-2.9.2 pydantic-core-2.23.4 sympy-1.13.1 torch-2.5.0 tqdm-4.66.5 typing-extensions-4.12.2
```

I just switched python to 3.11.Before I was using 3.10
```
conda create -n deepspeedcp311 python=3.11 pip
```
I read the document again, but there is no restriction on the Python version under Windows.
https://www.deepspeed.ai/tutorials/advanced-install/

**Comment by jhj0517** [comment]
@loadams Thanks for your work. Can you please upload whl for the new version as well?
I just reproduced this, so I tried to find the whl link for the new version of `deepspeed`, but it seems whl is only available for 0.15.1.
- https://pypi.org/project/deepspeed/#files

**Comment by loadams** [comment]
Hi @jhj0517 - yes we are working on getting updated whls published.  I'm hoping to have some by later this week for newer versions than 0.15.0/1.

**Comment by loadams** [comment]
@jhj0517 - we published whls for the latest, 0.16.3 for python 3.10, 3.11, and 3.12.  We will continue to publish going forward

**Comment by Romanio1997** [comment]
Downloading deepspeed-0.16.9.tar.gz (1.5 MB)
     ---------------------------------------- 1.5/1.5 MB 5.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [36 lines of output]
      test.c
      LINK : fatal error LNK1181: cannot open input file 'aio.lib'
      test.c
      LINK : fatal error LNK1181: cannot open input file 'cufile.lib'
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 35, in <module>
        File "C:\Users\user\AppData\Local\Temp\pip-install-mo75is1r\deepspeed_e24499fbdd5640e282a133313633ac0b\setup.py", line 201, in <module>
          ext_modules.append(builder.builder())
        File "C:\Users\user\AppData\Local\Temp\pip-install-mo75is1r\deepspeed_e24499fbdd5640e282a133313633ac0b\op_builder\builder.py", line 730, in builder
          extra_link_args=self.strip_empty_entries(self.extra_ldflags()))
        File "C:\Users\user\AppData\Local\Temp\pip-install-mo75is1r\deepspeed_e24499fbdd5640e282a133313633ac0b\op_builder\inference_cutlass_builder.py", line 74, in extra_ldflags
          import dskernels
      ModuleNotFoundError: No module named 'dskernels'
      DS_BUILD_OPS=1
       [WARNING]  Skip pre-compile of incompatible async_io; One can disable async_io with DS_BUILD_AIO=0
       [WARNING]  cpu_adam requires the 'lscpu' command, but it does not exist!
       [WARNING]  cpu_adam attempted to query 'lscpu' after failing to use py-cpuinfo to detect the CPU architecture. 'lscpu' does not appear to exist on your system, will fall back to use -march=native and non-vectorized execution.
       [WARNING]  cpu_adam requires the 'lscpu' command, but it does not exist!
       [WARNING]  cpu_adam attempted to query 'lscpu' after failing to use py-cpuinfo to detect the CPU architecture. 'lscpu' does not appear to exist on your system, will fall back to use -march=native and non-vectorized execution.
       [WARNING]  cpu_adagrad requires the 'lscpu' command, but it does not exist!
       [WARNING]  cpu_adagrad attempted to query 'lscpu' after failing to use py-cpuinfo to detect the CPU architecture. 'lscpu' does not appear to exist on your system, will fall back to use -march=native and non-vectorized execution.
       [WARNING]  cpu_adagrad requires the 'lscpu' command, but it does not exist!
       [WARNING]  cpu_adagrad attempted to query 'lscpu' after failing to use py-cpuinfo to detect the CPU architecture. 'lscpu' does not appear to exist on your system, will fall back to use -march=native and non-vectorized execution.
       [WARNING]  cpu_lion requires the 'lscpu' command, but it does not exist!
       [WARNING]  cpu_lion attempted to query 'lscpu' after failing to use py-cpuinfo to detect the CPU architecture. 'lscpu' does not appear to exist on your system, will fall back to use -march=native and non-vectorized execution.
       [WARNING]  cpu_lion requires the 'lscpu' command, but it does not exist!
       [WARNING]  cpu_lion attempted to query 'lscpu' after failing to use py-cpuinfo to detect the CPU architecture. 'lscpu' does not appear to exist on your system, will fall back to use -march=native and non-vectorized execution.
       [WARNING]  dc requires the 'lscpu' command, but it does not exist!
       [WARNING]  dc attempted to query 'lscpu' after failing to use py-cpuinfo to detect the CPU architecture. 'lscpu' does not appear to exist on your system, will fall back to use -march=native and non-vectorized execution.
       [WARNING]  dc requires the 'lscpu' command, but it does not exist!
       [WARNING]  dc attempted to query 'lscpu' after failing to use py-cpuinfo to detect the CPU architecture. 'lscpu' does not appear to exist on your system, will fall back to use -march=native and non-vectorized execution.
       [WARNING]  Skip pre-compile of incompatible evoformer_attn; One can disable evoformer_attn with DS_BUILD_EVOFORMER_ATTN=0
       [WARNING]  Skip pre-compile of incompatible fp_quantizer; One can disable fp_quantizer with DS_BUILD_FP_QUANTIZER=0
       [WARNING]  Skip pre-compile of incompatible gds; One can disable gds with DS_BUILD_GDS=0
       [WARNING]  Filtered compute capabilities [['6', '0'], ['6', '1'], ['7', '0']]
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

**Comment by loadams** [comment]
@Romanio1997 - we haven't published the whls yet for 0.16.9, we will work on this, but for now if you need to get a wheel for Windows you would need to download an older version.
